fix(console): add dynamic constraints layout in task details screen

vrn21-arcanist opened the pull request:
Changes I've made is for accommodating loong location names; ive changed the .. to the long name; it widens the task rectangle if its longer than the default (50%) and goes to the next line if its longer than what we can accommodate in a single line. This fixes issue #523  have attached screenshots of the same in the PR, before and after changes

These screenshots are what was the UI before my changes.
<img width="1440" alt="Screenshot 2025-03-16 at 4 15 10 PM" src="https://github.com/user-attachments/assets/c600b155-7b77-450e-9695-f9db9be67e8d" />


and these are the screenshots after my changes.
<img width="1440" alt="Screenshot 2025-03-16 at 4 14 25 PM" src="https://github.com/user-attachments/assets/83d1c6c6-5b06-42b8-a109-6e26d380eea2" />
<img width="1440" alt="Screenshot 2025-03-16 at 4 14 19 PM" src="https://github.com/user-attachments/assets/39d13ac2-de81-468c-b87a-2c9e52913108" />
<img width="1440" alt="Screenshot 2025-03-16 at 4 14 15 PM" src="https://github.com/user-attachments/assets/2bf01897-2d6a-4150-a245-a45bb3849543" />

vrn21-arcanist committed:
bug: dynamic alloc of layout constraints in task details screen

Changes i've made is for accomadating loong location names, ive changed the ... to the long name, it widens the task rectangle if its longer than the default 50% and goes to the next line if its longer than we can accomadate in a single line. This fixes issue #523
i have attached screenshots of the same in the PR, before and after changes

hds commented:
@vrn21 Thanks for working on this topic! So far this PR looks good. However, with a fixed number of rows, when you go to 2 lines for the location text, the "Idle" time is getting dropped off, so that needs to be resolved too. I think that there are some hints in the original issue regarding how to approach that issue.

vrn21-arcanist commented:
Thank you for looking into this @hds  and Sorry for the delay,

 but i'm still not sure how i'm supposed to move forward with this, i cant really determine what the height of the task_area is, before hand as in the current code, number of lines extra needed can be only computed after the each of the constraint is set. Could you help on this?

hds commented:
In the code that exists, the areas are computed up front, based on the constraints provided. However, those areas aren't actually used until the `Paragraph` for each of the task overview and the waker stats are actually rendered at the bottom.

If the creation of the task and waker "widgets" (that's the variable name), is performed before the area is split on constraints, then we can already compute the number of lines that each one needs and use that as input to the `Constraints::Length` for the `stats_area` which is where they sit.

Does that help?

vrn21-arcanist committed:
bug: idle line visible after dynamic alloc of layout constraints in task details screen

I've made the logic to look whether Location is longer than a single viewable length, even before setting constraints both vertically and horizontally. This fixes issue #523

vrn21-arcanist commented:
Sorry again for the delay, Does this work @hds?

I've changed the logic to check whether locations is longer than what is viewable in a single line, before hand, assigning constraints (both vertical and horizontal) afterwards.
<img width="1440" alt="Screenshot 2025-03-21 at 7 42 09 PM" src="https://github.com/user-attachments/assets/6c76f55c-4831-4767-b4d6-8d52fb7f7398" />
<img width="1440" alt="Screenshot 2025-03-21 at 7 41 27 PM" src="https://github.com/user-attachments/assets/c69bfc22-0192-4a1c-b162-f5f698d5fd3f" />

hds commented:
This is looking good! There are a few style and layout things, but the functionality appears correct. I'll get you a proper review of this on Monday.

Thanks for your work!

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -152,17 +247,13 @@ impl TaskView {
-        overview.push(Line::from(vec![bold(title), Span::raw(location)]));
+        let first_line = location_lines_vector[0].clone();
+        location_lines_vector.remove(0);
+        let location_vector = vec![bold(location_heading), Span::raw(first_line)];
```
I think that we can use the reference here, can't we? We don't need to clone or manipulate the Vec.

```suggestion
        let location_vector = vec![bold(location_heading), Span::raw(location_lines_vector[0])];
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -152,17 +247,13 @@ impl TaskView {
+        overview.push(Line::from(location_vector));
+        for line in location_lines_vector {
+            overview.push(Line::from(Span::raw(format!("    {}", line))));
+        }
```
Then here we can use the vec slice:

```suggestion
        for line in location_lines_vector[1..] {
            overview.push(Line::from(Span::raw(format!("    {}", line))));
        }
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -68,76 +67,172 @@ impl TaskView {
+
+        let location_heading = "Location: ";
+        let location_max_width = stats_area_check[0].width as usize - 2 - location_heading.len(); // NOTE: -2 for the border
+        let max_width_stats_area = area.width - 45;
```
Can we give the `45` a name? I'm guessing this is the minimum waker area needed?

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -68,76 +67,172 @@ impl TaskView {
+            .chunks(max_width_stats_area as usize)
+            .map(|chunk| chunk.iter().collect())
+            .collect();
+        let no_of_lines_extra_required_to_accomadate_location = location_lines_vector.len() - 1;
```
These very long variable names don't really fit in with the style of the rest of the codebase, we probably don't need the extra variable at all.

Instead, why don't we extract out the task stats height into a variable:

```suggestion
        // Id, Name, Target, Location (multiple), total, busy, scheduled, and idle times + top/bottom borders
        let task_stats_height = 9 + location_lines_vector.len();
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -68,76 +67,172 @@ impl TaskView {
-        ) = if warnings.is_empty() {
-            let chunks = Layout::default()
-                .direction(layout::Direction::Vertical)
+        ) = if task.location().len() > location_max_width {
```
We definitely don't want to add another layer of `if-else` statements here. I think that be specifying the task stats height outside, we can leave this like it was (just depending on the presense of warnings).

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -34,7 +34,6 @@ impl TaskView {
     pub(crate) fn update_input(&mut self, _event: input::Event) {
         // TODO :D
     }
-
```
I don't think we want to remove this line.

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -152,17 +247,13 @@ impl TaskView {
+        let location_vector = vec![bold(location_heading), Span::raw(first_line)];
+        overview.push(Line::from(location_vector));
+        for line in location_lines_vector {
+            overview.push(Line::from(Span::raw(format!("    {}", line))));
```
I think we need to add more spaces here so that the beginning of the text aligns with the first line location text.

If you leave only 4 spaces here, then the chunks for the 2nd line and onward should be bigger than the first chunk (which is also an option).

hds reviewed (changes requested):
This is looking good, thanks! There are some changes I've suggested.

vrn21-arcanist committed:
bug: dynamic alloc of layout constraints, requested changes changed

please refer to earlier commits to see what changes have been made

vrn21-arcanist commented:
@hds i've made the changes you've told to

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -68,6 +68,17 @@ impl TaskView {
+            .chunks(max_width_stats_area as usize)
+            .map(|chunk| chunk.iter().collect())
+            .collect();
+        // Id, Name, Target, Location (multiple), total, busy, scheduled, and idle times + top/bottom borders
```
I think I wrote something confusing, it's include the task stats height as a variable here:
```suggestion
        // Id, Name, Target, Location (multiple), total, busy, scheduled, and idle times + top/bottom borders
        let task_stats_height = 9 + location_lines_vector.len() as u16;
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -83,7 +94,7 @@ impl TaskView {
                         layout::Constraint::Length(controls.height()),
                         // task stats
-                        layout::Constraint::Length(10),
+                        layout::Constraint::Length((10 + location_lines_vector.len() - 1) as u16),
```
And use that variable here:

```suggestion
                        layout::Constraint::Length(task_stats_height),
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -105,7 +116,7 @@ impl TaskView {
                         layout::Constraint::Length(warnings.len() as u16 + 2),
                         // task stats
-                        layout::Constraint::Length(10),
+                        layout::Constraint::Length((10 + location_lines_vector.len() - 1) as u16),
```
And here too:

```suggestion
                        layout::Constraint::Length(task_stats_height),
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -127,16 +138,30 @@ impl TaskView {
+                    .as_ref(),
+                )
+                .split(stats_area)
+        };
```
We can simplify this a bit by just setting different constraints:

```suggestion
        let stats_constraints = if location_lines_vector.len() > 1 {
            [
                // 15 is the length of "| Location:     |"
                layout::Constraint::Min(area_needed_to_render_location + 15),
                layout::Constraint::Min(32),
            ]
        } else {
            [
                layout::Constraint::Percentage(50),
                layout::Constraint::Percentage(50),
            ]
        };
        let stats_area = Layout::default()
            .direction(layout::Direction::Horizontal)
            .constraints(stats_constraints.as_ref())
            .split(stats_area);
```

hds reviewed (changes requested):
One point here where I think I wasn't clear regarding what I meant. And one other thing that I didn't notice before, sorry about that.

vrn21-arcanist commented:
@hds i've updated the latest changes requested

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -127,15 +139,22 @@ impl TaskView {
+            let area_needed_to_render_location = task.location().len() as u16;
+            [
+                // 15 is the length of "| Location:     |"
+                layout::Constraint::Min(area_needed_to_render_location + 15),
```
Variable names are a bit long still. (you can hit "apply suggestion" to get the change directly applied)

```suggestion
            let location_len = task.location().len() as u16;
            [
                // 15 is the length of "| Location:     |"
                layout::Constraint::Min(location_len + 15),
```

hds commented on tokio-console/src/view/task.rs:
```diff
@@ -127,15 +139,22 @@ impl TaskView {
+            let area_needed_to_render_location = task.location().len() as u16;
+            [
+                // 15 is the length of "| Location:     |"
+                layout::Constraint::Min(area_needed_to_render_location + 15),
```
What happens when the 2 `Min` constraints add up to more than the available width?

vrn21 replied:
from [this](https://github.com/ratatui/ratatui/pull/393), i think the view would overflow according to the length the view has.

hds replied:
I think the `if` statement is incorrect then (and also back to front). In fact, I think that if instead the constraints look just like this then it might be enough (always, no if statement at all). Could you try and show some screenshots?

```rust
            [
                layout::Constraint::Min(location_lines_vector[0].len()),
                layout::Constraint::Min(32),
            ]
```

vrn21-arcanist committed:
bug fix: dynamic alloc of layout constraints in task details screen,

long var names changed to shorter names

Co-authored-by: Hayden Stainsby <hds@caffeineconcepts.com>

hds reviewed (commented):
What about this idea?

vrn21-arcanist commented:
Sorry for the delay again!

Yes, you're right, we could remove that unecessary if block. This doesnt really affect the view.

Attaching the current versions:

<img width="1440" alt="Screenshot 2025-03-28 at 12 22 20 AM" src="https://github.com/user-attachments/assets/f70bc890-eb99-46e6-87e9-dd48fad31ad6" />
<img width="1440" alt="Screenshot 2025-03-28 at 12 22 34 AM" src="https://github.com/user-attachments/assets/c46522e5-20f7-4804-9561-157aeb37f6f8" />
<img width="1440" alt="Screenshot 2025-03-28 at 12 22 46 AM" src="https://github.com/user-attachments/assets/9e04f5af-acde-4550-ae53-7185d257a398" />
<img width="1440" alt="Screenshot 2025-03-28 at 12 23 47 AM" src="https://github.com/user-attachments/assets/9a598a73-911a-40c1-9a6f-23986ad68fb3" />
<img width="1440" alt="Screenshot 2025-03-28 at 12 29 41 AM" src="https://github.com/user-attachments/assets/8cd6eaaa-db89-409b-a244-f1e9eeae992e" />
 
I'm commiting the changes( removing if block), i dont think this has any other problems with it.

vrn21-arcanist committed:
buf: refactored if block in task constraints

vrn21-arcanist commented:
@hds seems fine?

hds reviewed (approved):
Looks good. Thank you for all your work on this issue!
